SimG4CMS/Calo/src/CaloSteppingAction.cc

```diff
@@ -0,0 +1,427 @@
+#include "SimDataFormats/CaloHit/interface/PCaloHit.h"
+#include "Geometry/Records/interface/HcalSimNumberingRecord.h"
+
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/isFinite.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "G4LogicalVolumeStore.hh"
+#include "G4NavigationHistory.hh"
+#include "G4ParticleTable.hh"
+#include "G4PhysicalVolumeStore.hh"
+#include "G4RegionStore.hh"
+#include "G4Trap.hh"
+#include "G4UnitsTable.hh"
+#include "G4SystemOfUnits.hh"
+
+#include <cmath>
+#include <iostream>
+#include <iomanip>
+
+//#define EDM_ML_DEBUG
+
+CaloSteppingAction::CaloSteppingAction(const edm::ParameterSet &p) : 
+  count_(0) {
+
+  edm::ParameterSet iC = p.getParameter<edm::ParameterSet>("CaloSteppingAction");
+  nameEBSD_       = iC.getParameter<std::vector<std::string> >("EBSDNames");
+  nameEESD_       = iC.getParameter<std::vector<std::string> >("EESDNames");
+  nameHCSD_       = iC.getParameter<std::vector<std::string> >("HCSDNames");
+  nameHitC_       = iC.getParameter<std::vector<std::string> >("HitCollNames");
+  slopeLY_        = iC.getParameter<double>("SlopeLightYield");
+  birkC1EC_       = iC.getParameter<double>("BirkC1EC")*(g/(MeV*cm2));
+  birkSlopeEC_    = iC.getParameter<double>("BirkSlopeEC");
+  birkCutEC_      = iC.getParameter<double>("BirkCutEC");
+  birkC1HC_       = iC.getParameter<double>("BirkC1HC")*(g/(MeV*cm2));
+  birkC2HC_       = iC.getParameter<double>("BirkC2HC");
+  birkC3HC_       = iC.getParameter<double>("BirkC3HC");
+
+  edm::LogVerbatim("Step") << "CaloSteppingAction:: " << nameEBSD_.size() 
+			   << " names for EB SD's";
+  for (unsigned int k=0; k<nameEBSD_.size(); ++k)
+    edm::LogVerbatim("Step") << "[" << k << "] " << nameEBSD_[k];
+  edm::LogVerbatim("Step") << "CaloSteppingAction:: " << nameEESD_.size() 
+			   << " names for EE SD's";
+  for (unsigned int k=0; k<nameEESD_.size(); ++k)
+    edm::LogVerbatim("Step") << "[" << k << "] " << nameEESD_[k];
+  edm::LogVerbatim("Step") << "CaloSteppingAction:: " << nameHCSD_.size() 
+			   << " names for HC SD's";
+  for (unsigned int k=0; k<nameHCSD_.size(); ++k)
+    edm::LogVerbatim("Step") << "[" << k << "] " << nameHCSD_[k];
+  edm::LogVerbatim("Step") << "CaloSteppingAction::Constants for ECAL: slope "
+			   << slopeLY_ << " Birk constants " << birkC1EC_ 
+			   << ":" << birkSlopeEC_ << ":" << birkCutEC_;
+  edm::LogVerbatim("Step") << "CaloSteppingAction::Constants for HCAL: Birk "
+			   << "constants " << birkC1HC_ << ":" << birkC2HC_
+			   << ":" << birkC3HC_;
+  edm::LogVerbatim("Step") << "CaloSteppingAction:: " << nameHitC_.size() 
+			   << " hit collections";
+  for (unsigned int k=0; k<nameHitC_.size(); ++k)
+    edm::LogVerbatim("Step") << "[" << k << "] " << nameHitC_[k];
+
+  ebNumberingScheme_.reset(new EcalBarrelNumberingScheme());
```

Comment:
better to use `ebNumberingScheme_ = std::make_unique<EcalBarrelNumberingScheme>();` and similar for all `unique_ptr`s below

Reply:
Done